### ISSUE TITLE
Prevent loosing jobs by using the Reliable Queue Pattern

### ISSUE THREAD

**Issue by thanpolas** [verbling/kickq#4]
Kickq uses the [Blocking Pop (BLPOP)](http://redis.io/commands/blpop) command to fetch jobs that are ready to be processed. If a client fails or disconnects, the job will perish forever as it has been removed from the queue.

This behavior results in some tests failing randomly. The case is that a previous test, may have the time to re-send a _BLPOP_ command before it gets disposed. This results in the next test to fail as the job will be popped by the previous, now disposed, worker.

An interim solution would be to find a way to send an _abort_ command to the _blocked client_, but it appears that this cannot be done explicitly.

A test case has been written that illuminates this weakness and awaits for a passing solution.

---

TL;DR

The best long term solution is to enforce the [_Reliable Queue Pattern_](http://redis.io/commands/rpoplpush) by using the [BRPOPLPUSH](http://redis.io/commands/brpoplpush) command.

This requires the following:
- Switch from [RPUSH](http://redis.io/commands/rpush) to [LPUSH](http://redis.io/commands/lpush) to retain FIFO when using _BRPOPLPUSH_.
- When a Job completes, with any outcome, the [LREM](http://redis.io/commands/lrem) command must be used to remove the Job Id from the _Processing_ queue.
- Create a guard that monitors the _Processing_ queue.
### The Guard

That's a pickle. 
#### Draft Solution One, Polling

The Guard will poll the _Processing_ queue and store in an Array all the Job Ids that are being processed, along with the timestamp of when they were first seen.

A configurable timeout (~2s) will trigger the _Inspector_ for Job Ids that have surpassed that limit. The _Inspector_ will fetch the _Job Item_ and examine it. If the _Job Item_ checks out then the [Processing Timeout property](https://github.com/verbling/kickq/wiki/Configure-Kickq#option--processtimeout) is read and a timer is created with a +20% margin.

If the Job's Processing Timeout +20% is surpassed and the Job Id is still in the _Processing_ queue then it will be moved back to the _To Process_ queue.

**cons**: Too much noise, by having all the workers polling the _Processing_ queue and fetching Job Items that surpass the ~2s timeout.
## Help

I am not very happy with that solution. 

Please share your thoughts.
